rwkv.cpp

```diff
@@ -629,326 +989,197 @@ struct rwkv_context * rwkv_init_from_file(const char * file_path, const uint32_t
+    size_t max_out_size = 0;
+    size_t max_key_length = 0;
 
-        RWKV_ASSERT_FALSE(RWKV_ERROR_MODEL_PARAMS, read_int32(file_in, &key_length, "key_length"));
-        RWKV_ASSERT_FALSE(RWKV_ERROR_MODEL_PARAMS, read_int32(file_in, &parameter_data_type, "parameter_data_type"));
+    while (ftell(in_file) < in_stat.st_size) {
+        struct tensor_header header;
+        RWKV_ASSERT_FALSE(RWKV_ERROR_FILE, fread_tensor_header_and_skip(in_file, header));
 
-        RWKV_ASSERT_FALSE_MSG(RWKV_ERROR_MODEL_PARAMS | RWKV_ERROR_SHAPE, n_dims == 1 || n_dims == 2, "Unsupported dimension count %d", n_dims);
-        RWKV_ASSERT_FALSE_MSG(
-            RWKV_ERROR_MODEL_PARAMS | RWKV_ERROR_UNSUPPORTED,
-            parameter_data_type >= 0 && parameter_data_type < FORMAT_TYPE_COUNT,
-            "Unsupported parameter data type %d",
-            parameter_data_type
-        );
+        size_t in_size = tensor_bytes(header);
+        if (in_size > max_in_size) max_in_size = in_size;
 
-        ggml_type parameter_ggml_type = FORMAT_TYPE_TO_GGML_TYPE[parameter_data_type];
-        RWKV_ASSERT_FALSE_MSG(
-            RWKV_ERROR_MODEL_PARAMS | RWKV_ERROR_UNSUPPORTED,
-            parameter_ggml_type != GGML_TYPE_UNKNOWN,
-            "Unsupported parameter data type %d",
-            parameter_data_type
-        );
+        // f16 type tensors get relocated to out and then converted into f32 at in
+        if (header.data_type == TYPE_F16) {
+            if (in_size > max_out_size) max_out_size = in_size;
+            size_t f32_size = tensor_bytes(GGML_TYPE_F32, header.width, header.height);
+            if (f32_size > max_in_size) max_in_size = f32_size;
+        }
 
-        int32_t nelements, x, y;
+        size_t out_size = tensor_bytes(out_type, header.width, header.height);
+        if (out_size > max_out_size) max_out_size = out_size;
 
-        if (n_dims == 1) {
-            RWKV_ASSERT_FALSE(RWKV_ERROR_MODEL_PARAMS, read_int32(file_in, &x, "x"));
-            y = 1;
-            nelements = x;
-        } else {
-            RWKV_ASSERT_FALSE(RWKV_ERROR_MODEL_PARAMS, read_int32(file_in, &x, "x"));
-            RWKV_ASSERT_FALSE(RWKV_ERROR_MODEL_PARAMS, read_int32(file_in, &y, "y"));
-            nelements = x * y;
-        }
+        if (header.key_length > max_key_length) max_key_length = header.key_length;
+    }
 
-        std::string name(key_length, 0);
-        RWKV_ASSERT_NULL_MSG(RWKV_ERROR_MODEL_PARAMS | RWKV_ERROR_KEY, fread(&name[0], key_length, 1, file_in) == 1, "Failed to read parameter key");
-
-        printf("%48s - [%5d, %5d], type = %6s ", name.data(), x, y, ggml_type_name(parameter_ggml_type));
-        total_size_orig += (size_t) (nelements * ggml_type_sizef(parameter_ggml_type));
-
-        // Quantize only 2D tensors, except embedding and head matrices.
-        // Embedding and head take not too much space, especially in bigger models;
-        // but they significantly increase perplexity when quantized.
```

Comment:
I propose to keep this comment for some additional clarity

Reply:
sorry about that, I didn't remove this on purpose, just forgot to add it back